server/graphql/types/src/types/repack.rs

```diff
@@ -0,0 +1,145 @@
+use async_graphql::{dataloader::DataLoader, *};
+use chrono::NaiveDateTime;
+use graphql_core::{
+    loader::{LocationByIdLoader, StockLineByIdLoader},
+    standard_graphql_error::StandardGraphqlError,
+    ContextExt,
+};
+use service::repack::query::Repack;
+
+use super::{LocationNode, StockLineNode};
+
+pub struct RepackNode {
+    // Invoice id
+    pub id: String,
+    pub repack_id: String,
+    pub batch: Option<String>,
+    pub datetime: NaiveDateTime,
+    pub from: RepackStockLineNode,
+    pub to: RepackStockLineNode,
+}
+
+pub struct RepackStockLineNode {
+    // Repacked number of packs before conversion (for from stock line)
+    pub number_of_packs: f64,
+    pub pack_size: i32,
+    pub location_id: Option<String>,
+    pub stock_line_id: Option<String>,
+}
+
+#[Object]
+impl RepackNode {
+    async fn id(&self) -> &str {
+        &self.id
+    }
+
+    async fn repack_id(&self) -> &str {
+        &self.repack_id
+    }
+
+    async fn batch(&self) -> &Option<String> {
+        &self.batch
+    }
+
+    async fn datetime(&self) -> NaiveDateTime {
+        self.datetime
+    }
+
+    async fn from(&self) -> &RepackStockLineNode {
+        &self.from
+    }
+
+    async fn to(&self) -> &RepackStockLineNode {
+        &self.to
+    }
+}
+
+#[Object]
+impl RepackStockLineNode {
+    async fn number_of_packs(&self) -> f64 {
+        self.number_of_packs
+    }
+
+    async fn pack_size(&self) -> i32 {
+        self.pack_size
+    }
+
+    async fn location(&self, ctx: &Context<'_>) -> Result<Option<LocationNode>> {
+        let loader = ctx.get_loader::<DataLoader<LocationByIdLoader>>();
+
+        let location_id = match &self.location_id {
+            None => return Ok(None),
+            Some(location_id) => location_id,
+        };
+
+        let result = loader.load_one(location_id.clone()).await?;
+
+        Ok(result.map(LocationNode::from_domain))
+    }
+
+    async fn stock_line(&self, ctx: &Context<'_>) -> Result<Option<StockLineNode>> {
+        if let Some(stock_line_id) = &self.stock_line_id {
+            let loader = ctx.get_loader::<DataLoader<StockLineByIdLoader>>();
+            let stock_line = loader.load_one(stock_line_id.clone()).await?.ok_or(
+                StandardGraphqlError::InternalError(format!(
+                    "Cannot find stock line {}",
+                    stock_line_id
+                ))
+                .extend(),
+            )?;
+            Ok(Some(StockLineNode { stock_line }))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+impl RepackNode {
+    pub fn from_domain(repack: Repack) -> RepackNode {
+        RepackNode {
```

Comment:
I think it would look slightly better if you extra invoice_row, to_invoice_line_row and from_invoice_line_row first, before the mapping

Reply:
[3e02fca](https://github.com/openmsupply/open-msupply/pull/1796/commits/3e02fcae2776ca47c033a933761c437607593c58)